datanode/snapshot/database_meta_data.go

```diff
@@ -0,0 +1,167 @@
+package snapshot
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"code.vegaprotocol.io/vega/datanode/sqlstore"
+	"github.com/georgysavva/scany/pgxscan"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/jackc/pgx/v4/stdlib"
+	"github.com/pressly/goose/v3"
+)
+
+type DatabaseMetadata struct {
+	TableNameToMetaData map[string]TableMetadata
+	DatabaseVersion     int64
+}
+
+type TableMetadata struct {
+	Name       string
+	SortOrder  string
+	Hypertable bool
+}
+
+type IndexInfo struct {
+	Tablename string
+	Indexname string
+	Indexdef  string
+}
+
+func NewDatabaseMetaData(ctx context.Context, connConfig sqlstore.ConnectionConfig) (DatabaseMetadata, error) {
+	conn, err := pgxpool.Connect(ctx, connConfig.GetConnectionString())
+	if err != nil {
+		return DatabaseMetadata{}, fmt.Errorf("unable to connect to database: %w", err)
+	}
+	defer conn.Close()
+
+	tableNames, err := GetAllTableNames(ctx, conn)
+	if err != nil {
+		return DatabaseMetadata{}, fmt.Errorf("failed to get names of tables to copy:%w", err)
+	}
+
+	tableNameToSortOrder, err := getTableSortOrders(ctx, conn)
+	if err != nil {
+		return DatabaseMetadata{}, fmt.Errorf("failed to get table sort orders:%w", err)
+	}
+
+	hyperTableNames, err := getHyperTableNames(ctx, conn)
+	if err != nil {
+		return DatabaseMetadata{}, fmt.Errorf("failed to get hyper table names:%w", err)
+	}
+
+	dbVersion, err := getDatabaseVersion(connConfig)
+	if err != nil {
+		return DatabaseMetadata{}, fmt.Errorf("failed to get database version:%w", err)
+	}
+
+	result := DatabaseMetadata{TableNameToMetaData: map[string]TableMetadata{}, DatabaseVersion: dbVersion}
+	for _, tableName := range tableNames {
+		result.TableNameToMetaData[tableName] = TableMetadata{
+			Name:       tableName,
+			SortOrder:  tableNameToSortOrder[tableName],
+			Hypertable: hyperTableNames[tableName],
+		}
+	}
+
+	return result, nil
+}
+
+func (d DatabaseMetadata) GetHistoryTableNames() []string {
+	var result []string
+	for _, meta := range d.TableNameToMetaData {
+		if meta.Hypertable {
+			result = append(result, meta.Name)
+		}
+	}
+
+	return result
+}
+
+func getDatabaseVersion(connConfig sqlstore.ConnectionConfig) (int64, error) {
+	poolConfig, err := connConfig.GetPoolConfig()
+	if err != nil {
+		return 0, fmt.Errorf("failed to get pool config:%w", err)
+	}
+
+	db := stdlib.OpenDB(*poolConfig.ConnConfig)
+	defer db.Close()
+
+	dbVersion, err := goose.GetDBVersion(db)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get goose database version:%w", err)
+	}
+	return dbVersion, nil
+}
+
+func GetAllTableNames(ctx context.Context, conn *pgxpool.Pool) ([]string, error) {
+	tableNameRows, err := conn.Query(ctx, "SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' and table_type = 'BASE TABLE' and"+
+		" table_name != 'goose_db_version' order by table_name")
```

Comment:
why is that 2 string concatenated here?

Reply:
to split across the IDE, my monitors not that wide :)  do we care?